opentracing/propagation.py

```diff
@@ -0,0 +1,169 @@
+
+        :return: a Span instance joined to the trace state in `carrier` or None
+            if no such trace state could be found.
+        """
+        raise NotImplementedError()
+
+
+class InvalidCarrierException(Exception):
+    """InvalidCarrierException should be used when the provided carrier
+    instance does not match what the `format` argument requires.
+
+    See Injector and Extractor.
+    """
+    pass
+
+
+class TraceCorruptedException(Exception):
+    """TraceCorruptedException should be used when the underlynig trace state
+    is seemingly present but not well-formed.
+
+    See Extractor.join_trace.
+    """
+    pass
+
+
+class Format:
+    """A namespace for builtin Injector/Extractor formats.
+
+    These static constants are intended for use in the Tracer.injector() and
+    Tracer.extractor() methods. E.g.,
+
+        tracer.injector(Format.SPLIT_BINARY).inject_span(...)
+
+    """
+
+    # The SPLIT_BINARY format pairs with an Injector or Extractor that expects
+    # a SplitBinaryCarrier carrier object.
+    SPLIT_BINARY = 'split_binary'
+
+    # The SPLIT_TEXT format pairs with an Injector or Extractor that expects a
+    # SplitTextCarrier carrier object.
+    SPLIT_TEXT = 'split_text'
+
+
+class SplitBinaryCarrier(object):
+    """The SplitBinaryCarrier is a carrier to be used with Format.SPLIT_BINARY
+    Injectors/Extractors.
+
+    The SplitBinaryCarrier has two properties, and each is represented as a
+    bytearray:
+     - tracer_state: Tracer-specific context that must cross process
+       boundaries. For example, in Dapper this would include a trace_id, a
+       span_id, and a bitmask representing the sampling status for the given
+       trace.
+     - trace_attributes: any Trace Attributes for the encoded Span (per
+       Span.SetTraceAttribute).
+    """
+
+    def __init__(self, tracer_state=None, trace_attributes=None):
+        self.tracer_state = (
+            bytearray() if tracer_state is None else tracer_state)
+        self.trace_attributes = (
+            bytearray() if trace_attributes is None else trace_attributes)
+
+
+class SplitTextCarrier(object):
+    """The SplitTextCarrier is a carrier to be used with Format.SPLIT_BINARY
+    Injectors/Extractors.
+
+    The SplitTextCarrier has two properties, and each is represented as a
+    string->string dict:
+     - tracer_state: Tracer-specific context that must cross process
+       boundaries. For example, in Dapper this would include a trace_id, a
+       span_id, and a bitmask representing the sampling status for the given
+       trace.
+     - trace_attributes: any Trace Attributes for the encoded Span (per
+       Span.SetTraceAttribute).
+    """
+
+    def __init__(self, tracer_state=None, trace_attributes=None):
+        self.tracer_state = (
+            {} if tracer_state is None else tracer_state)
+        self.trace_attributes = (
+            {} if trace_attributes is None else trace_attributes)
+
+
+class _NoopPropagator:
```

Comment:
@yurishkuro yes, exactly... I don't mind private knowledge of various classes within a single python package, hence the direction I went here. It's a compromise between this ugliness and an optional `tracer=None` arg in the `Extractor.__init__` method... Since the only use case I had in mind was this noop impl detail, I didn't want to make changes to the public API. If you care about this, I'm happy to change it, though, def a minor point in my book.